### PR TITLE
[SPARK-18746][SQL] Add implicit encoder for BigDecimal, timestamp and date

### PR DESCRIPTION
**Pull request by weiqingy**
## What changes were proposed in this pull request?
Add implicit encoders for BigDecimal, timestamp and date. 

## How was this patch tested?
Add an unit test.  Pass build, unit tests, and some tests below .
Before:
```
scala> spark.createDataset(Seq(new java.math.BigDecimal(10)))
<console>:24: error: Unable to find encoder for type stored in a Dataset.  Primitive types (Int, String, etc) and Product types (case classes) are supported by importing spark.implicits._  Support for serializing other types will be added in future releases.
       spark.createDataset(Seq(new java.math.BigDecimal(10)))
                          ^

scala>
```
After:
```
scala> spark.createDataset(Seq(new java.math.BigDecimal(10)))
res0: org.apache.spark.sql.Dataset[java.math.BigDecimal] = [value: decimal(38,18)]
```

### CODE REVIEW

**Comment by weiqingy** [comment]
@gatorsmile Could you please review this PR? Thanks.

**Comment by gatorsmile** [comment]
cc @cloud-fan

**Comment by weiqingy** [comment]
The only failure is “[org.apache.spark.sql.kafka010.KafkaSourceStressForDontFailOnDataLossSuite.stress test for failOnDataLoss=false](https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceSuite.scala#L848)”, but it passed locally. I'll re-trigger again.

**Comment by weiqingy** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala [hunk 1]
```diff
@@ -74,6 +74,9 @@ abstract class SQLImplicits {
   /** @since 1.6.0 */
   implicit def newStringEncoder: Encoder[String] = Encoders.STRING
 
+  /** @since 2.2.0 */
+  implicit def newDecimalEncoder: Encoder[java.math.BigDecimal] = Encoders.DECIMAL
+
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: shall we add all of them? i.e. timestamp, date, binary encoder.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: @cloud-fan, do we need to add Scala's `scala.math.BigDecimal` one too? I am just curious.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: yea we should

Comment (weiqingy) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala]: Thanks for the feedback, @cloud-fan @HyukjinKwon. I will update the PR to add more encoders.

**Comment by weiqingy** [comment]
Hi, @cloud-fan  I have updated the PR to add more encoders. Could you please review this PR again? Thanks.

**Comment by cloud-fan** [comment]
sorry it conflict... can you update?

**Comment by weiqingy** [comment]
Yes, I have updated the PR. @cloud-fan

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by weiqingy** [comment]
Thanks for the review. @cloud-fan

**Pull request closed without merging**